crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,148 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
         })
         .collect()
 }
+
+/// Semaphore like implementation that allows both shrinking and expanding
+/// the max permits.
+#[derive(Clone, Debug)]
+pub(crate) struct ResizableSemaphore(Arc<SemShared>);
```

Comment:
Please group impls together with their definitions rather than having all definitions and then all impls. This is the convention we have everywhere else and makes code discovery a bit easier.